### DISCUSSION TITLE
getInitialProps vs getStaticProps

### DISCUSSION THREAD

**Discussion by vasanthi-d** [vercel/next.js#40124] [Help] [answered]
Next.js docs say recommended approach is getStaticProps or getServerSideProps

https://nextjs.org/docs/api-reference/data-fetching/get-initial-props

Then, my understanding is that we can either use getStaticProps or getServerSideProps and not bother about getInitialProps for SSR.

What is the difference between getInitialProps and  getStaticProps / getServerSideProps?
Are there any use cases where getInitialProps should be used instead of getStaticProps / getServerSideProps ?

**Comment by icyJoseph** [comment, marked as answer]
`GetServerSideProps` (GSSP) vs `getInitialProps` is the real question here, because once you activate either, for a `page`, then the page is rendered on demand.

`getStaticProps`, without `ISR` taken into account, simply fetches props when `next build` is executed and passes that to its associated page. That's not something `getInitialProps` can accomplish. 

Then when it comes to `GSSP` vs `getInitialProps`, there's a fundamental difference, and that is the quirky nature of `getInitialProps`.

`GSSP` strictly runs server side, whereas, `getInitialProps`, runs as you navigate into a page, but also when the is prefetched, which means that if you have logic dependent on `req`, or `res` objects, on the client you have to check for their presence.

There's a bit of a rabbit hole here with `next/link`, prefetching and `getInitialProps` during SPA transitions, which I do not want to explore here, but look at it like this. Using the https://next-code-elimination.vercel.app/, we can see that:

```js
import Cookies from 'cookies';
import Mysql from 'mysql';
import Link from 'next/link';
import SQL from 'sql-template-strings';
import Layout from '../components/Layout';

const pool = Mysql.createPool(process.env.DATABASE_URL);

export const MyPage = ({ projects }) => {
  return (
    <Layout>
      <h1>Projects</h1>
      <ul>
        {projects.map((project) => (
          <li key={project.id}>
            <Link href="/projects/[id]" as={`/projects/${project.id}`}>
              <a>{project.name}</a>
            </Link>
          </li>
        ))}
      </ul>
    </Layout>
  );
}

MyPage.getInitialProps = async function getInitialProps(ctx) {
  const userId = new Cookies(req, res).get('user_id');
  const projects = await new Promise((resolve, reject) =>
    pool.query(
      SQL`SELECT id, name FROM projects WHERE user_id = ${userId};`,
      (err, results) => (err ? reject(err) : resolve(results))
    )
  );
  return { projects };
}
```

is bundled as:

```js
// This is the code that is bundled for the client-side:

import Cookies from 'cookies';
import Mysql from 'mysql';
import Link from 'next/link';
import SQL from 'sql-template-strings';
import Layout from '../components/Layout';
const pool = Mysql.createPool(process.env.DATABASE_URL);
export const MyPage = ({ projects }) => {
  return (
    <Layout>
      <h1>Projects</h1>
      <ul>
        {projects.map((project) => (
          <li key={project.id}>
            <Link href="/projects/[id]" as={`/projects/${project.id}`}>
              <a>{project.name}</a>
            </Link>
          </li>
        ))}
      </ul>
    </Layout>
  );
};
MyPage.getInitialProps = async function getInitialProps(ctx) {
  const userId = new Cookies(req, res).get('user_id');
  const projects = await new Promise((resolve, reject) =>
    pool.query(
      SQL`SELECT id, name FROM projects WHERE user_id = ${userId};`,
      (err, results) => (err ? reject(err) : resolve(results))
    )
  );
  return { projects  };
};
```

Which is already problematic! We are leaking code to the client, moreover, we are trying to import the MySql library to create pools of connections, and so on. Not good.

Take a look at the same but using `getServerSideProps`

```js
import Cookies from 'cookies';
import Mysql from 'mysql';
import Link from 'next/link';
import SQL from 'sql-template-strings';
import Layout from '../components/Layout';

const pool = Mysql.createPool(process.env.DATABASE_URL);

export const MyPage = ({ projects }) => {
  return (
    <Layout>
      <h1>Projects</h1>
      <ul>
        {projects.map((project) => (
          <li key={project.id}>
            <Link href="/projects/[id]" as={`/projects/${project.id}`}>
              <a>{project.name}</a>
            </Link>
          </li>
        ))}
      </ul>
    </Layout>
  );
}

export async function getServerSideProps({ req, res }) {
  const userId = new Cookies(req, res).get('user_id');
  const projects = await new Promise((resolve, reject) =>
    pool.query(
      SQL`SELECT id, name FROM projects WHERE user_id = ${userId};`,
      (err, results) => (err ? reject(err) : resolve(results))
    )
  );
  return { props: { projects } };
}
```

Then the client bundle:

```js
// This is the code that is bundled for the client-side:

import Link from 'next/link';
import Layout from '../components/Layout';
export var __N_SSP = true;
export const MyPage = ({ projects }) => {
  return (
    <Layout>
      <h1>Projects</h1>
      <ul>
        {projects.map((project) => (
          <li key={project.id}>
            <Link href="/projects/[id]" as={`/projects/${project.id}`}>
              <a>{project.name}</a>
            </Link>
          </li>
        ))}
      </ul>
    </Layout>
  );
};
```

Much better! We've made it so that the server side is strictly taken away. 

> Note that it is possible to get to somewhat the same with getInitialProps, but it is not trivial, https://arunoda.me/blog/ssr-and-server-only-modules

`GetServerSideProps` and `GetStaticProps` are relatively new, added back in `9.3.x`, and they were designed to cater to issues seen by the community when working with `getInitialProps`. Personally, having worked with next before v9, was not so pleasant, but I was quickly drawn back in because of these two additions.

Today the most outstanding use case is `getInitialProps` in `_app`, which has its own caveats, but its necessity comes from the need to fetch data on the server, at runtime, for the entire app, not just a page. At build time, `_app` getInitialProps, runs once per generated page (pre-rendered).

**Reply by vasanthi-d** [reply]
Thank you @icyJoseph for clear and detailed answer.   I feel better now!!
I think this explanation should be added to docs.

is getInitialProps in `_app` supported in Next 12.1.x?   If so, can you share any example?  

Thank you for the great explanation.
